### ISSUE TITLE
Support for S3

### ISSUE THREAD

**Issue by IljaN** [owncloud/data_exporter#27]
- [ ] Should technically work, need smoke-test
- [ ] Add tests

**Comment by jvillafanez** [comment]
Require full test
[ ] from local storage to S3
[ ] from S3 to local storage

That should also cover S3 to S3

**Comment by IljaN** [comment]
- [ ] Create custom import process for versions in S3, as S3 native versioning is used.

**Comment by PVince81** [comment]
Maybe https://github.com/owncloud/core/blob/v10.0.10/lib/public/Files/ObjectStore/IVersionedObjectStorage.php#L77 after getting the storage instance and checking whether it implements this interface.

The interface and impl of S3 would need to be adjusted to also accept the mtime.

As @DeepDiver1975 mentioned today, there is a risk that the underlying native S3 versioning API doesn't support setting specific mtimes...

**Comment by jvillafanez** [comment]
Unless there is a stablished public API in core to handle versions, I'd rather not to use anything depending on the implementation, otherwise the behaviour will be different depending on the type of storage we're using.

For now, versions will be created by uploading content to the same location. The mtime will be different, but that's a known limitation regardless of the storage.
If the retention policy is different depending on the storage, that's out of the app's scope

**Comment by IljaN** [comment]
> Unless there is a stablished public API in core to handle versions, I'd rather not to use anything 
> depending on the implementation...

You are correct, by pushing files in to files_version directory we are also already using an implementation detail which happens to work.
